source/d3d8to9_device.cpp

```diff
@@ -2134,3 +2173,18 @@ void Direct3DDevice8::ApplyClipPlanes()
 		index++;
 	}
 }
+
+void Direct3DDevice8::ReleaseShaders()
+{
+	for (auto Handle : PixelShaderHandles)
+	{
+		DeletePixelShader(Handle);
+	}
+	PixelShaderHandles.clear();
+	for (auto Handle : VertexShaderHandles)
+	{
+		DeleteVertexShader(Handle);
+	}
+	VertexShaderHandles.clear();
+	VertexShaderAndDeclarationCount = 0;
+}
```

Comment:
This does not reset all the values for the shaders.  d3d8to9 also caches shaders so we should reset the cache like this:

```c++
CurrentVertexShaderHandle = 0;
CurrentPixelShaderHandle = 0;
```

Also, I am not sure if Direct3D9 remembers which shader was set before the Release and if we should reset that also?

```c++
ProxyInterface->SetPixelShader(nullptr);
ProxyInterface->SetVertexShader(nullptr);
```

Reply:
This is handled by the calls to `Delete(Pixel/Vertex)Shader` (which will call `Set(Pixel/Vertex)Shader(0)` for the respective current handle once encountered). So it should actually reset those values already.